### PR TITLE
Use concurrent dictionary iso reader writer lock slim

### PR DESCRIPTION
**Pull request by FrankBakkerNl**

### CODE REVIEW

File: Src/FluidCaching/Index.cs [hunk 1]
```diff
@@ -69,11 +67,12 @@ public void Remove(TKey key)
...
         {
-            return RWLock.GetReadLock(_lock, LockTimeout, delegate
-            {
-                WeakReference value;
-                return (INode<T>) (index.TryGetValue(key, out value) ? value.Target : null);
-            });
+            WeakReference<INode<T>> reference;
+            INode<T> node;
+            if (index.TryGetValue(key, out reference) && reference.TryGetTarget(out node))
+                return node;
```
Comment (dennisdoomen) [inline Src/FluidCaching/Index.cs:74]: Use parentheses after an `if`.

File: Src/FluidCaching/Index.cs [hunk 2]
```diff
@@ -111,16 +101,11 @@ public int RebuildIndex()
...
-                    index.Clear();
-                    foreach (INode<T> item in lifespanManager)
-                    {
-                        AddItem(item);
-                    }
-
-                    return index.Count;
-                });
+                // Create a new ConcurrentDictionary, this way there is no need for locking the index itself
+                var keyValues = lifespanManager.Select(item => new KeyValuePair<TKey, WeakReference<INode<T>>>(_getKey(item.Value), new WeakReference<INode<T>>(item)));
```
Comment (dennisdoomen) [inline Src/FluidCaching/Index.cs:107]: Can't you use `ToDictionary` instead?

**Comment by dennisdoomen** [comment]
:+1:

File: Src/FluidCaching/Index.cs (hunk 2 above)
Comment (FrankBakkerNl) [inline Src/FluidCaching/Index.cs:107]: We would need a need a ToConcurentDictionary() for that, which is not available in the framework. Implementing it here for single use seems a bit overkill.

I do agree this code is bit complex, mainly due to all the nested generics (imagine if i had not used var here :-) something like IEnumerable<KeyValuePair<Tkey, WeakReference<INode<T>>>

**Pull request merged**